### PR TITLE
[singlejar] Port test_util to Windows

### PR DESCRIPTION
**Pull request by rongjiecomputer**
- Port various functions in `test_util` to Windows.
- Replace `@local_jdk//:jdk-default` with `@bazel_tools//tools/jdk:current_java_runtime`.

See #2241

### CODE REVIEW

**Comment by rongjiecomputer** [comment]
PTAL @meteorcloudy

**Comment by meteorcloudy** [comment]
I've imported this change, but the C++ runfiles library doesn't work well internally. I need some time to debug and fix the issue. Sorry for the delay.

**Comment by laszlocsomor** [comment]
I'll take a look at why the runfiles library is failing in the Google-internal code base.

File: src/tools/singlejar/BUILD [hunk 1]
```diff
@@ -125,7 +125,7 @@ cc_test(
         ":input_jar",
         ":test_util",
         "//src/main/cpp/util",
-        "@com_google_googletest//:gtest_main",
+        "@com_google_googletest//:gtest",
```
Comment (laszlocsomor) [inline src/tools/singlejar/BUILD]: Please revert changes in this file. See my other comments for reason.

File: src/tools/singlejar/BUILD [hunk 2]
```diff
@@ -415,7 +416,8 @@ cc_library(
     hdrs = ["test_util.h"],
     deps = [
         "//src/main/cpp/util",
-        "@com_google_googletest//:gtest_main",
+        "@bazel_tools//tools/cpp/runfiles",
```
Comment (laszlocsomor) [inline src/tools/singlejar/BUILD]: Please, for now, depend on `//tools/cpp/runfiles`.
I'm about to send a patch to the runfiles library that'll fix the issue with the Google-internal use case.
Please also add a TODO with my name (e.g. `TODO(laszlocsomor)`) saying I should use this library from `@bazel_tools`.

File: src/tools/singlejar/test_util.cc [hunk 3]
```diff
@@ -114,4 +135,19 @@ string CreateTextFile(const string& relpath, const char *contents) {
...
 }
 
+Runfiles* runfiles = nullptr;
+
 }  // namespace singlejar_test_util
+
+int main(int argc, char** argv) {
+  std::string error;
+  std::unique_ptr<singlejar_test_util::Runfiles> runfiles(
+      singlejar_test_util::Runfiles::Create(argv[0], &error));
```
Comment (laszlocsomor) [inline src/tools/singlejar/test_util.cc]: Unfortunately I can't fix the error without patching the runfiles library, which I'm about to do. Could you please hold this PR until that one is merged?

My patch will, among other things, add a static `Runfiles* Runfiles::CreateForTest(void)` method, which won't require `argv0`. Therefore there'll be no need for a custom `main` method.

Once my patch is in, could you please init the `runfiles` either on-demand (in tests that need it) or in the `setUp` method, using the upcoming `CreateForTest` factory method?

File: src/tools/singlejar/test_util.cc [hunk 4]
```diff
@@ -114,4 +135,19 @@ string CreateTextFile(const string& relpath, const char *contents) {
   return string("");
 }
 
+Runfiles* runfiles = nullptr;
```
Comment (laszlocsomor) [inline src/tools/singlejar/test_util.cc]: Is there a way to not use a global variable? (Maybe by initializing `runfiles` in the `setUp` method?)

Comment (rongjiecomputer) [reply src/tools/singlejar/test_util.cc]: Initializing `runfiles` in the `setUp` is fine. I used global variable when I started porting tests because I did not know how much I need to call `runfiles->Rlocation`.

File: src/tools/singlejar/test_util.cc (hunk 3 above)
Comment (rongjiecomputer) [reply src/tools/singlejar/test_util.cc]: > Could you please hold this PR until that one is merged?

Sure thing. Ping me when that PR is merged. #6251 probably will need some changes too, so don't review that one yet.

**Comment by rongjiecomputer** [comment]
@laszlocsomor This PR is ready now.

**Comment by laszlocsomor** [comment]
Thanks @rongjiecomputer ! I don't see the code using the runfiles library anywhere. Did you mean to use it, or just adding the dependency?

**Comment by rongjiecomputer** [comment]
That will be in #6251, which I have not yet updated.

File: src/tools/singlejar/test_util.cc [hunk 5]
```diff
@@ -87,7 +108,7 @@ string GetEntryContents(const string &zip_path, const string &entry_name) {
   string command;
   blaze_util::StringPrintf(&command, "unzip -p %s %s", zip_path.c_str(),
                            entry_name.c_str());
-  FILE *fp = popen(command.c_str(), "r");
+  FILE *fp = popen(command.c_str(), "rb");
```
Comment (laszlocsomor) [inline src/tools/singlejar/test_util.cc]: Unfortunately this change breaks the internal version of `output_jar_simple_test`.
Is this change necessary?

Comment (laszlocsomor) [reply src/tools/singlejar/test_util.cc]: http://man7.org/linux/man-pages/man3/popen.3.html doesn't document `b`; it seems to be Microsoft-specific.

Comment (rongjiecomputer) [reply src/tools/singlejar/test_util.cc]: The `"b"` is necessary because otherwise the Microsoft implementation will strip `\r\n` to `\n` and make many tests fail. I will add a `_WIN32` guard for this.

Comment (laszlocsomor) [reply src/tools/singlejar/test_util.cc]: Thanks!
With your permission, I'd like to also add this comment:
```
  // "b" is Microsoft-specific. It's necessary, because without it the popen
  // implementation will strip "\r\n" to "\n" and make many tests fail.
```
(If you agree to it, just let me know, no need to upload a new commit.)

Comment (rongjiecomputer) [reply src/tools/singlejar/test_util.cc]: You have my permission.

File: src/tools/singlejar/BUILD [hunk 6]
```diff
@@ -232,12 +232,13 @@ cc_test(
         ":test1",
         ":test2",
         "@local_jdk//:jar",
-        "@local_jdk//:jdk-default",
+        "@local_jdk//:jdk",
```
Comment (ulfjack) [inline src/tools/singlejar/BUILD]: Ideally, this would be fixed not to use @local_jdk. Also, did this test not actually run before? There's no :jdk-default target, I think.

Comment (rongjiecomputer) [reply src/tools/singlejar/BUILD]: > Ideally, this would be fixed not to use `@local_jdk`

So what is the alternative?

I also have been wondering whether the singlejar tests were even tested before, and whether the `//src/tools/singlejar/BUILD` in Bazel is different from the Google internal one (because some test files are never included in the `BUILD` file and see this comment: https://github.com/bazelbuild/bazel/pull/6336#discussion_r223815429).

Comment (laszlocsomor) [reply src/tools/singlejar/BUILD]: I believe `//external:jdk` and `//external:jar` would be more desirable. Alas, I'm not at all familiar with these JDK rules. I'm consulting with @lberki for advice on the internal review.

Comment (lberki) [reply src/tools/singlejar/BUILD]: Me! If all the test needs is to run `jar`, do this instead:

```
cc_test(
  name="test",
  data=["@bazel_tols//tools/jdk:current_java_runtime"],
  toolchains=["@bazel_tools//tools/jdk:current_java_runtime"],
  copts=["-DJAR=$(JAVA)/bin/jar"],
)
```

This way, you get a path to whatever target JDK is used (and not to the result of autodetection), which is the principled thing to do.

Comment (rongjiecomputer) [reply src/tools/singlejar/BUILD]: I think we need to add `.exe` at for the `-D` flag. Windows also requires different kind of `"` escaping to pass string constant in `-D` as `const char[]`. I will handle that in the second PR, just do whatever that works for Unix first.

Comment (lberki) [reply src/tools/singlejar/BUILD]: Sounds like a plan. You can use `select({"@bazel_tools//platforms:windows": ... })` or something like that to handle that case.

Comment (laszlocsomor) [reply src/tools/singlejar/BUILD]: @rongjiecomputer : shall I apply @lberki 's suggestion (the cc_test rule) to the code and submit internally like so?

Comment (rongjiecomputer) [reply src/tools/singlejar/BUILD]: Sure.
